### PR TITLE
iCal Support

### PR DESCRIPTION
**Pull request by kjoconnor**
This replaces https://github.com/midgetspy/Sick-Beard/pull/179 and renders a string directly as you asked.  Also generally cleaner than the last version.

### CODE REVIEW

**Comment by akuiraz** [comment]
Where do I go to get my iCal file with this pull request?  (I've already merged your branch with my copy of Sick-Beard...)

**Comment by kjoconnor** [comment]
Should be at /ical

Sent from my iPhone

On Sep 11, 2012, at 2:54, akuiraz notifications@github.com wrote:

Where do I go to get my iCal file with this pull request? (I've already
merged your branch with my copy of Sick-Beard...)

—
Reply to this email directly or view it on
GitHubhttps://github.com/midgetspy/Sick-Beard/pull/464#issuecomment-8448772.

**Comment by akuiraz** [comment]
I get this error... http://pastebin.com/VmtWzf1d

**Comment by Ronner** [comment]
I hope midgetspy merges this soon as I would LOVE to finally be able to have iCal support in sickbeard.

kjoconner, do you have some instructions on how to use this with my current sickbeard install? I'm a bit of a git newbie and have no clue how I would add this to my current sickbeard install (on ubuntu server).

**Comment by Ronner** [comment]
I made a clone of https://github.com/kjoconnor/Sick-Beard.git and checked out the ical-support branch to give it a try. Then copied my settings and database over and got it to work.

However, when I go to http://username:password@mydomain:8081/ical, it works, but when I use that url in google calendar to add it as a calendar, it runs into an error that it can not open the url. I have no clue whether it's google's servers that can't retrieve it, or that the result they receive is simply not valid.

Have you tried to add an ical feed from sickbeard as a google calendar?

**Comment by kjoconnor** [comment]
I have, it's probably because your installation is not web-accessible I'd
guess.  Go to web-sniffer.net and put in your full iCalendar URL that
you're trying to add in Google Calendar and see if it resolves properly.

**Comment by Ronner** [comment]
My installation is web-accessible. I have tons of things running on my homeserver that is accessible from the web, including sickbeard.

I opted for something different instead. Wrote a php script that uses the sickbeard api (future command with today|soon|later) and outputs it in iCalender format, and used the url to that php script in google calendar. Works like a charm.

**Comment by akuiraz** [comment]
Care to share your PHP script?  Would love to try it out... @Ronner

**Comment by kjoconnor** [comment]
@akuiraz that error is probably due to a show not having a proper `air_time`.  Can you check if that's the case?

**Comment by Ronner** [comment]
@akuiraz Sure, no prob. 

``` php
<?php
$username = "YourUsername";
$password = "YourPassword";
$ip = "127.0.0.1:8081";
$api = "YourAPIkey";

// Change the date interval to your needs. For instance, in Europe sickbeard 
// usually downloads the show the next day so I want to see it in my calendar
// on the day that it is downloaded, because that is the day that I can actually
// watch it.
$interval = new DateInterval('P1D');

if($username) {
    $feed = "http://".$username.":".$password."@".$ip."/api/".$api."/?cmd=future&sort=date&type=today|soon|later";
} else {
    $feed = "http://".$ip."/api/".$api."/?cmd=future&sort=date&type=today|soon|later";
}

$sbJSON = json_decode(file_get_contents($feed));

echo "BEGIN:VCALENDAR\r\n";
echo "X-ORIGINAL-URL:http://www.sickbeard.com\r\n";
echo "X-WR-CALNAME:Upcoming episodes\r\n";
echo "VERSION:2.0\r\n";
echo "METHOD:PUBLISH\r\n";

foreach($sbJSON->{data} as $section) {
        foreach($section as $show) {

            $date = date_create_from_format( 'Y-m-d', $show->{airdate});
            $date->add($interval);
            $outDate = $date->format('Ymd');

            echo "BEGIN:VEVENT\r\n";
            echo "SUMMARY:" . $show->{show_name} . " S" . sprintf("%02s", $show->{season}) . "E" . sprintf("%02s", $show->{episode}) . "\r\n";
            echo "UID:http://www.sickbeard.com#" . $show->{show_name} . " S" . sprintf("%02s", $show->{season}) . "E" . sprintf("%02s", $show->{episode}) . "\r\n";
            echo "DTSTART;VALUE=DATE:" . $outDate . "\r\n";
            echo "DTEND;VALUE=DATE:" . $outDate . "\r\n";
            echo "DTSTAMP:" . $outDate . "T060000Z" . "\r\n";
            echo "END:VEVENT\r\n";
        }
}

echo "END:VCALENDAR\r\n"
?>
```

**Comment by akuiraz** [comment]
@kjoconnor Yes I think this may be it, what is the format needed by this?

**Comment by lrusak** [comment]
it would be nice to see this rebased and merged

**Comment by kjoconnor** [comment]
@lrusak unfortunately 'base' seems to be a moving target, this is the fourth time I've rebased it to not have it accepted

**Comment by kjoconnor** [comment]
Back again, @midgetspy / @thezoggy does this look good?

**Comment by mbnn** [comment]
Guys, how is this supposed to work I tried adding urltosickbeard:8081/ical to google but i keep getting error. What is the exact URL to add?

**Comment by kjoconnor** [comment]
This is really really far out of date so I doubt it works anymore.

File: sickbeard/webserve.py [hunk 1]
```diff
@@ -2828,6 +2830,98 @@ def comingEpisodes(self, layout="None"):
...
+        except ImportError:
+            pil_import = False
+
+        poster_url = cherrypy.url().replace('ical', '')
+
+        time_re = re.compile('([0-9]{1,2})\:([0-9]{2})(\ |)([AM|am|PM|pm]{2})')
+
+        cal = Calendar()
+
+        # As far as I know Sick-Beard is pinned to EST
```
Comment (thezoggy) [inline sickbeard/webserve.py:2852]: sb itself uses whatever timezone the user is. the data we pull down (tvrage/tvdb) does not have a timezone or can be trusted to have one.

Comment (kjoconnor) [inline sickbeard/webserve.py:2852]: This PR is over a year old so I don't imagine it's good anymore.  I closed it.

**Pull request closed without merging**